### PR TITLE
[Refactor] Refactor code structure and docstrings

### PR DESCRIPTION
**Pull request by THU17cyz**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
Please describe the motivation of this PR and the goal you want to achieve through this PR.

There are some confusing or unnecessary or erroneous code:
- The naming of `points_in_bbox` and `points_in_bbox_batch`
- The same code in `CameraInstance3DBoxes`, `LidarInstance3DBoxes`, and `DepthInstance3DBoxes`
- Error in the conversion between depth boxes and cam boxes

There are also some docstrings need fixing:
- Some functions are for cam boxes only/Some functions are for lidar/depth boxes only, but lack the notation
- The docstrings in `mmdet3d/core/post_processing/box3d_nms.py` and `mmdet3d/ops/iou3d/iou3d_utils.py`
- Missing 'optional' tag for function parameters, and wrong/missing/mispelled 'default value notation'
- Some params with the option of `None` have 'None' in the docstring (e.g. `(torch.Tensor | None)`), but most don't. MMDetection also show the same trend. For consistency I removed all such occurences of 'None'. 
- 'w, l, h' are used instead of 'x_size, y_size, z_size', or 'dx, dy, dz' are used instead 'x_size, y_size, z_size' (may be confused with delta x,y,z in mmdet3d/core/bbox/coders/delta_xyzwhlr_bbox_coder.py)
- Useless `\`s in docstrings

## Modification
Please briefly describe what modification is made in this PR.
Code refactor:
- Renaming of `points_in_bbox` and `points_in_bbox_batch` to `points_in_bbox_part` and `points_in_bbox_all`
- Merging some methods of `CameraInstance3DBoxes`, `LidarInstance3DBoxes`, and `DepthInstance3DBoxes`
- Fix the error in box conversion (and added unittests)

Docstring fix:
- Modifying docstring for `points_in_bbox_part` and `points_in_bbox_all`
- Rewrite docstrings in `mmdet3d/core/post_processing/box3d_nms.py` and `mmdet3d/ops/iou3d/iou3d_utils.py`
- Adding notation of the type of boxes appliable to box-processing functions
- Adding 'optional' tag for function parameters, and fixing wrong/missing/mispelled 'default value notation' (through searching with regex in VSCode). More specificly, `Defaults to xxx` and `Default: xxx` have no grammar errors. (If Defaults to ... is not present, I will not add the optional tag, since for some functions all default related docstring is missing and the workload for this job is too big)
- Removing all 'None's in the function param types in the docstring
- Replacing 'w, l, h' with 'x_size, y_size, z_size', and replacing misused 'dx, dy, dz' with 'x_size, y_size, z_size'
- Remove useless `\`s in docstrings (used regex to fix all occurences + manual check)

## BC-breaking (Optional)
Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)
If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by THU17cyz** [comment]
Will add some unittests to make sure everything is still correct. May review first.

File: mmdet3d/models/dense_heads/anchor3d_head.py [hunk 1]
```diff
@@ -317,7 +317,7 @@ class predictions.
                 of each sample.
             gt_labels (list[torch.Tensor]): Gt labels of each sample.
             input_metas (list[dict]): Contain pcd and img's meta info.
-            gt_bboxes_ignore (None | list[torch.Tensor]): Specify
+            gt_bboxes_ignore (list[torch.Tensor]): Specify
                 which bounding.
```
Comment (Wuziyi616) [inline mmdet3d/models/dense_heads/anchor3d_head.py:321]: `Specify which bounding box to ignore`?

File: mmdet3d/models/roi_heads/h3d_roi_head.py [hunk 2]
```diff
@@ -67,11 +67,11 @@ def forward_train(self,
...
             gt_labels_3d (list[torch.Tensor]): Labels of each sample.
-            pts_semantic_mask (None | list[torch.Tensor]): Point-wise
+            pts_semantic_mask (list[torch.Tensor]): Point-wise
                 semantic mask.
-            pts_instance_mask (None | list[torch.Tensor]): Point-wise
+            pts_instance_mask (list[torch.Tensor]): Point-wise
                 instance mask.
-            gt_bboxes_ignore (None | list[torch.Tensor]): Specify
+            gt_bboxes_ignore (list[torch.Tensor]): Specify
                 which bounding.
```
Comment (Wuziyi616) [inline mmdet3d/models/roi_heads/h3d_roi_head.py:75]: `which bounding box to ignore`?

File: mmdet3d/ops/iou3d/iou3d_utils.py [hunk 3]
```diff
@@ -22,24 +22,29 @@ def boxes_iou_bev(boxes_a, boxes_b):
...
+    two boxes. In this function, one can also set `pre_max_size` and
+    `post_max_size`.
 
     Args:
         boxes (torch.Tensor): Input boxes with the shape of [N, 5]
             ([x1, y1, x2, y2, ry]).
         scores (torch.Tensor): Scores of boxes with the shape of [N].
         thresh (int): Threshold.
-        pre_maxsize (int): Max size of boxes before nms. Default: None.
-        post_maxsize (int): Max size of boxes after nms. Default: None.
```
Comment (Wuziyi616) [inline mmdet3d/ops/iou3d/iou3d_utils.py:34]: `post_max_size`

File: mmdet3d/ops/roiaware_pool3d/src/points_in_boxes_cuda.cu [hunk 4]
```diff
@@ -74,11 +74,11 @@ __global__ void points_in_boxes_kernel(int batch_size, int boxes_num,
   }
 }
 
-__global__ void points_in_boxes_batch_kernel(int batch_size, int boxes_num,
+__global__ void points_in_boxes_all_kernel(int batch_size, int boxes_num,
                                              int pts_num, const float *boxes,
                                              const float *pts,
                                              int *box_idx_of_points) {
```
Comment (Wuziyi616) [inline mmdet3d/ops/roiaware_pool3d/src/points_in_boxes_cuda.cu:80]: Fix the args indent here.

File: mmdet3d/ops/roiaware_pool3d/src/points_in_boxes_cuda.cu [hunk 5]
```diff
@@ -127,17 +127,17 @@ void points_in_boxes_launcher(int batch_size, int boxes_num, int pts_num,
 #endif
 }
 
-void points_in_boxes_batch_launcher(int batch_size, int boxes_num, int pts_num,
+void points_in_boxes_all_launcher(int batch_size, int boxes_num, int pts_num,
                                     const float *boxes, const float *pts,
                                     int *box_idx_of_points) {
```
Comment (Wuziyi616) [inline mmdet3d/ops/roiaware_pool3d/src/points_in_boxes_cuda.cu:132]: indent

File: mmdet3d/ops/roiaware_pool3d/src/points_in_boxes_cuda.cu [hunk 6]
```diff
@@ -151,9 +151,9 @@ void points_in_boxes_batch_launcher(int batch_size, int boxes_num, int pts_num,
 #endif
 }
 
-int points_in_boxes_gpu(at::Tensor boxes_tensor, at::Tensor pts_tensor,
+int points_in_boxes_part(at::Tensor boxes_tensor, at::Tensor pts_tensor,
                         at::Tensor box_idx_of_points_tensor) {
```
Comment (Wuziyi616) [inline mmdet3d/ops/roiaware_pool3d/src/points_in_boxes_cuda.cu:155]: indent

File: mmdet3d/ops/roiaware_pool3d/src/points_in_boxes_cuda.cu [hunk 7]
```diff
@@ -176,9 +176,9 @@ int points_in_boxes_gpu(at::Tensor boxes_tensor, at::Tensor pts_tensor,
   return 1;
 }
 
-int points_in_boxes_batch(at::Tensor boxes_tensor, at::Tensor pts_tensor,
+int points_in_boxes_all(at::Tensor boxes_tensor, at::Tensor pts_tensor,
                           at::Tensor box_idx_of_points_tensor) {
```
Comment (Wuziyi616) [inline mmdet3d/ops/roiaware_pool3d/src/points_in_boxes_cuda.cu:180]: indent

File: mmdet3d/ops/roiaware_pool3d/src/points_in_boxes_cuda.cu [hunk 8]
```diff
@@ -194,7 +194,7 @@ int points_in_boxes_batch(at::Tensor boxes_tensor, at::Tensor pts_tensor,
   const float *pts = pts_tensor.data_ptr<float>();
   int *box_idx_of_points = box_idx_of_points_tensor.data_ptr<int>();
 
-  points_in_boxes_batch_launcher(batch_size, boxes_num, pts_num, boxes, pts,
+  points_in_boxes_all_launcher(batch_size, boxes_num, pts_num, boxes, pts,
                                  box_idx_of_points);
```
Comment (Wuziyi616) [inline mmdet3d/ops/roiaware_pool3d/src/points_in_boxes_cuda.cu:198]: indent

File: mmdet3d/ops/roiaware_pool3d/src/roiaware_pool3d.cpp [hunk 9]
```diff
@@ -40,16 +40,16 @@ int roiaware_pool3d_gpu_backward(at::Tensor pts_idx_of_voxels,
 int points_in_boxes_cpu(at::Tensor boxes_tensor, at::Tensor pts_tensor,
                         at::Tensor pts_indices_tensor);
 
-int points_in_boxes_gpu(at::Tensor boxes_tensor, at::Tensor pts_tensor,
+int points_in_boxes_part(at::Tensor boxes_tensor, at::Tensor pts_tensor,
                         at::Tensor box_idx_of_points_tensor);
```
Comment (Wuziyi616) [inline mmdet3d/ops/roiaware_pool3d/src/roiaware_pool3d.cpp:44]: indent

File: mmdet3d/ops/roiaware_pool3d/src/roiaware_pool3d.cpp [hunk 10]
```diff
@@ -40,16 +40,16 @@ int roiaware_pool3d_gpu_backward(at::Tensor pts_idx_of_voxels,
...
 
-int points_in_boxes_batch(at::Tensor boxes_tensor, at::Tensor pts_tensor,
+int points_in_boxes_all(at::Tensor boxes_tensor, at::Tensor pts_tensor,
                           at::Tensor box_idx_of_points_tensor);
```
Comment (Wuziyi616) [inline mmdet3d/ops/roiaware_pool3d/src/roiaware_pool3d.cpp:47]: indent

**Review by Wuziyi616 — COMMENTED** [review]
I assume you have tested the correctness of these modifications. I only point out some minor comments

File: mmdet3d/core/bbox/structures/base_box3d.py [hunk 11]
```diff
@@ -460,34 +506,43 @@ def new_box(self, data):
...
+            points (torch.Tensor): Points in shape (1, M, 3) or (M, 3),
+                3 dimensions are (x, y, z) in LiDAR or depth coordinate.
+            boxes_override (torch.Tensor, optional): Boxes to override
+                `self.tensor `. Defaults to None.
 
         Returns:
-            torch.Tensor: The index of box where each point are in.
+            torch.Tensor: The index of the box in which
+                each point is, in shape (M, ). Default value is -1
+                (if the point is not enclosed by any box).
```
[the remark this answers is not in the document]
Comment (THU17cyz) [reply mmdet3d/core/bbox/structures/base_box3d.py:521]: Sure. Somehow I forgot to fix the return description. My fault.

File: mmdet3d/core/bbox/structures/base_box3d.py [hunk 12]
```diff
@@ -460,34 +506,43 @@ def new_box(self, data):
...
+            points (torch.Tensor): Points in shape (1, M, 3) or (M, 3),
+                3 dimensions are (x, y, z) in LiDAR or depth coordinate.
+            boxes_override (torch.Tensor, optional): Boxes to override
+                `self.tensor `. Defaults to None.
 
         Returns:
-            torch.Tensor: The index of boxes each point lies in with shape
-                of (B, M, T).
+            torch.Tensor: The index of all boxes in which each point is,
+                in shape (B, M, T).
```
[the remark this answers is not in the document]
Comment (THU17cyz) [reply mmdet3d/core/bbox/structures/base_box3d.py]: T is num_boxes (convention, see https://github.com/open-mmlab/mmdetection3d/mmdet3d/ops/roiaware_pool3d/points_in_boxes.py) Just rewrote the docstring. Great idea!

File: mmdet3d/ops/roiaware_pool3d/src/points_in_boxes_cuda.cu [hunk 13]
```diff
@@ -102,19 +102,19 @@ __global__ void points_in_boxes_batch_kernel(int batch_size, int boxes_num,
   }
 }
 
-void points_in_boxes_launcher(int batch_size, int boxes_num, int pts_num,
+void points_in_boxes_part_launcher(int batch_size, int boxes_num, int pts_num,
                               const float *boxes, const float *pts,
                               int *box_idx_of_points) {
```
Comment (Wuziyi616) [inline mmdet3d/ops/roiaware_pool3d/src/points_in_boxes_cuda.cu:107]: indent

**Review by Wuziyi616 — APPROVED** [review]
Almost ready to merge.

File: mmdet3d/core/bbox/structures/base_box3d.py [hunk 14]
```diff
@@ -460,34 +506,49 @@ def new_box(self, data):
         return original_type(
             new_tensor, box_dim=self.box_dim, with_yaw=self.with_yaw)
 
-    def points_in_boxes(self, points, boxes_override=None):
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/structures/base_box3d.py:463]: deprecated warning.

File: mmdet3d/core/bbox/structures/base_box3d.py [hunk 15]
```diff
@@ -460,34 +506,49 @@ def new_box(self, data):
...
+        if points.dim() == 2:
+            points = points.unsqueeze(0)
+        box_idx = points_in_boxes_part(points,
+                                       boxes.unsqueeze(0).to(
+                                           points.device)).squeeze(0)
         return box_idx
 
-    def points_in_boxes_batch(self, points, boxes_override=None):
-        """Find points that are in boxes (CUDA).
+    def points_in_boxes_all(self, points, boxes_override=None):
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/structures/base_box3d.py:538]: deprecated warning

**Pull request merged**